### ISSUE TITLE
Noob question - How to Filter a Page grid queries different for different users

### ISSUE THREAD

**Issue by brunobola** [volkanceylan/Serenity#1082]
Noob question - How to Filter a Page grid queries different for different users.

PS. I'm not a web developer i used to be a winforms developer so yes, i'll make noob questions ;)

**Comment by SimeonDD** [comment]
I guess you mean to check the login credentials of the user and then apply a custom filter on the page.
If that is the case I do not think Serene offers the functionality out of the box (pending confirmation from the forum).

Serene, in a nutshell, offers a great starting page which is instantly useful. It shrinks 2-3 days of coding to about 1 hr. However for any customization beyond that point you need to be absolutely fluent in TypeScript and JavaScript (lots of ECMAScript 6 notations). In addition you need to be familiar with how the blocks in an MVC based project work. Then you will be able to reuse the Serene libraries to develop your new code.

Not trying to put you down or anything cause I am exactly in your position.
I am myself moving away from the world of windows programming and it is difficult to not be able to "just add a button" in a line or two. Here the code is in a lot of places and there are no shortcuts. Just a completely different way of thinking.

Basically - if you are fluent in this programming Serene is a great time saver.
If you are not - you can use Serene to poke around and teach yourself, but you will absolutely not be able to quickly spin out a full-blown production grade custom application.

**Comment by brunobola** [comment]
Humm, i'm a fast learner and with a few code and samples i believe i can do it, just need a little guidance.
by the way, your tutorials are excellent.

Thanks.

**Comment by SimeonDD** [comment]
Oh ... man. Not my tutorials. It is all Volkan.
I am new to this field like you.

A possible workaround would be to figure out how to populate a page from a stored procedure and pass the user to the SP as a parameter.
There were some posts to that effect on the forum.
Again - to the best of my knowledge - all that will require custom programming.

About "some code samples". That is exactly what Serene is actually. A gigantic code sample. But that ain't helping much if you are not completely fluent in reading it.

Not putting Serene down. But it is like a you have been a drummer and somebody gives you that great guitar with all the sheet-music ... you ain't playing that thing for a while for sure.

**Comment by brunobola** [comment]
Sorry, my bad, of course they are all volkan, quickly response just in a middle of trying something's, sorry.

I know mvc, asp, etc, but came from a devexpress usability environment and try to make a shift, and learn a bit.

Melhores cumprimentos.
Eng° Bruno Marques

Enviado por TypeApp



Em 19 de ago de 2016, 22:10, em 22:10, SimeonDD <notifications@github.com> escreveu:

**Comment by mlutzow** [comment]
With the help of volkan, after I had added an additional field to the users table, I added the below code in my ...repository.cs files that were of interest for me.  You would be doing a little something different, but I think it's the same concept.

        private class MyListHandler : ListRequestHandler<MyRow>
        {
            protected override void ApplyFilters(SqlQuery query)
            {
                base.ApplyFilters(query);

                var user = (UserDefinition)Authorization.UserDefinition;

                query.Where(fld.Division == user.Division);
`

**Comment by SimeonDD** [comment]
2nd Workaround (no stored proc)
Serene's page already comes with two types of filters. The default one on top and if you want columns specific filters.

If you figure out how to read the loginID and pass it to the code of the page then you could potentially write a custom rule that will always pre-filter the page using its already available filters.

**Comment by brunobola** [comment]
Thanks, for the code share, that's what i need to brick some walls.

**Comment by biggikalli** [comment]
Hi guys,

I had the same problem recently on how to filter the grid, only showing the current's user data in the grid and not the all the table data. Sort of a semi-semi multi-tenancy.

I used and tweaked this from a xxxGrid.ts in the Basic Samples and it workes perfectly:

`

        protected onViewSubmit() {
            // only continue if base class returns true (didn't cancel request)
            if (!super.onViewSubmit()) {
                return false;
            }

            // view object is the data source for grid (SlickRemoteView)
            // this is an EntityGrid so its Params object is a ListRequest
            var request = this.view.params as Serenity.ListRequest;

            // list request has a Criteria parameter
            // we AND criteria here to existing one because 
            // otherwise we might clear filter set by 
            // an edit filter dialog if any.

            request.Criteria = Serenity.Criteria.and(request.Criteria,
                [['Username'], '=', PerDiem.Authorization.userDefinition.Username]);

            // TypeScript doesn't support operator overloading
            // so we had to use array syntax above to build criteria.

            // Make sure you write
            // [['Field'], '>', 10] (which means field A is greater than 10)
            // not 
            // ['A', '>', 10] (which means string 'A' is greater than 10

            return true;
        }`

The grid only shows the data belonging to the logged in user by filtering the username directly into the grid. 

I am also new to web developing, having several winforms (VB!!) programs under the belt. However, being with Serenity through the Saltarella era and also now the Typescript, I can surely say that you don't have to be a member of Mensa to put out professional applications with Serenity. That is exactly what it is for, cutting down production time by MASSIVE quantities of hours. When you become familiar with it you can create a fully blown system with the fraction of the time needed to make a prototype from the ground up.

I myself have fired up several very different systems built on Serenity and I just love it.

Please don't give up on this platform easily though it's learning curve might look steep. All the samples , tutorials and information in issues on forums give immense support to a starting developer in Serenity.

I fully recommend it.

**Comment by volkanceylan** [comment]
I recommend doing filtering on server side, at repository, if security is a concern. Otherwise @biggikali method is also fine. You can access user info with Authorization.UserDefinition both in TypeScript/CS

**Comment by SimeonDD** [comment]
I just found the solutions to issue #128 which will allow you to use a stored procedure.
This is also a viable approach.

**Comment by ahmedabdelh** [comment]
You can use the session
